File: frontend/src/api/gddNotionSync.ts
/**
 * Client API synchronisation GDD depuis Notion (FR18).
 */
import { API_TIMEOUTS } from '../constants'
import apiClient from './client'

/** Clé React Query pour l’état checkpoint / reprise sync complète. */
export const GDD_FULL_SYNC_CHECKPOINT_QUERY_KEY = ['gdd-full-sync-checkpoint'] as const

export interface GddNotionSyncConfigPublic {
  schema_version: number
  sync_interval_minutes: number
  auto_sync_enabled: boolean
  sources: Array<{
    notion_id: string
    kind: 'database' | 'page'
    category_file: string
    /** Bases multi-vues : forcer les UUID data source Notion (vide = auto côté serveur). */
    notion_data_source_ids?: string[]
  }>
  included_categories: string[]
  mirror_rebuild_on_full_sync: boolean
  archive_retention_count: number
  token_configured: boolean
}

export interface GddNotionSyncConfigResponse {
  config: GddNotionSyncConfigPublic
}

/** Corps PUT /config (tous les champs optionnels côté API). */
export interface GddNotionSyncConfigUpdateBody {
  sync_interval_minutes?: number
  auto_sync_enabled?: boolean
  sources?: GddNotionSyncConfigPublic['sources']
  included_categories?: string[]
  mirror_rebuild_on_full_sync?: boolean
  archive_retention_count?: number
  notion_token?: string
}

export interface GddNotionConnectionTestResponse {
  ok: boolean
  message: string
  bot_id?: string
  bot_type?: string
}

export interface GddNotionSyncRunResponse {
  success: boolean
  message: string
  updated_entities: number
  partial_errors: string[]
  mirror_promotion_pending?: boolean
}

export interface GddNotionSyncStatusResponse {
  last_started_at: string | null
  last_finished_at: string | null
  last_success: boolean | null
  message: string
  updated_entities: number
  partial_errors: string[]
  last_archive_relative?: string | null
  last_mirror_rebuild_used?: boolean | null
  mirror_promotion_pending?: boolean | null
}

export interface GddNotionSyncProgressResponse {
  active: boolean
  started_at: string | null
  force_full: boolean | null
  mirror_rebuild: boolean | null
  phase: string
  sources_total: number
  sources_completed: number
  current_source_index: number
  current_category_file: string
  pages_total_known: number
  pages_processed: number
  pages_in_current_source: number
  current_page_in_source: number
  current_page_id_short: string
  message: string
  paused?: boolean
}

export interface GddFullSyncCheckpointResponse {
  resumable: boolean
  /** none | resumable | stale | invalid_file */
  checkpoint_status: string
  checkpoint_file_present: boolean
  orphan_staging_runs: number
  message: string
  staging_run_name: string
  archive_rel: string
  sources_total: number
  sources_completed: number
  completed_category_files: string[]
  eligible_category_files: string[]
  mirror_promotion_pending?: boolean
}

export interface GddFullSyncSimpleOkResponse {
  ok: boolean
  message: string
}

export async function getGddNotionSyncConfig(): Promise<GddNotionSyncConfigResponse> {
  const { data } = await apiClient.get<GddNotionSyncConfigResponse>(
    '/api/v1/gdd-notion-sync/config',
  )
  return data
}

export async function putGddNotionSyncConfig(
  body: GddNotionSyncConfigUpdateBody,
): Promise<GddNotionSyncConfigResponse> {
  const { data } = await apiClient.put<GddNotionSyncConfigResponse>(
    '/api/v1/gdd-notion-sync/config',
    body,
  )
  return data
}

export async function postGddNotionTestConnection(): Promise<GddNotionConnectionTestResponse> {
  const { data } = await apiClient.post<GddNotionConnectionTestResponse>(
    '/api/v1/gdd-notion-sync/test-connection',
  )
  return data
}

/** Réponse POST prévisualisation première ligne d’une base (alignée Pydantic). */
export interface GddNotionPreviewDatabaseResponse {
  ok: boolean
  message: string
  category_file: string
  notion_database_id: string
  data_sources_count: number
  data_source_entries: Array<{ id: string; name: string }>
  query_total_rows: number
  first_page_id: string
  property_keys_from_query_row: string[]
  property_keys_from_get_page: string[]
  mapped_record: Record<string, unknown> | null
  compact_table: boolean
}

/** Une ligne comme en sync : utile pour vérifier Notion avant une sync complète. */
export async function postGddNotionPreviewDatabaseRow(
  categoryFile: string,
): Promise<GddNotionPreviewDatabaseResponse> {
  const { data } = await apiClient.post<GddNotionPreviewDatabaseResponse>(
    '/api/v1/gdd-notion-sync/preview-database-row',
    { category_file: categoryFile },
    { timeout: API_TIMEOUTS.LLM_GENERATION },
  )
  return data
}

export interface PostGddNotionSyncOptions {
  resume?: boolean
  fresh?: boolean
  /** Limite ce run aux ``category_file`` listés (bases) ; sans suppression des autres bases. */
  categoryFiles?: string[]
  /** Applique le staging conservé après erreurs partielles (full=true requis). */
  applyStagingDespiteErrors?: boolean
  /** Filtre UI éphémère (non persisté). ``[]`` = toutes les bases cochées / pas de filtre. */
  includedCategories?: string[]
}

export async function postGddNotionSync(
  full = false,
  opts?: PostGddNotionSyncOptions,
): Promise<GddNotionSyncRunResponse> {
  /** FastAPI attend ``category_file=a&category_file=b`` ; Axios mettrait ``category_file[]`` (ignoré). */
  const search = new URLSearchParams()
  search.set('full', full ? 'true' : 'false')
  if (opts?.resume) {
    search.set('resume', 'true')
  }
  if (opts?.fresh) {
    search.set('fresh', 'true')
  }
  if (opts?.applyStagingDespiteErrors) {
    search.set('apply_staging_despite_errors', 'true')
  }
  if (opts?.includedCategories !== undefined) {
    search.set('included_filter', 'ui')
    for (const raw of opts.includedCategories) {
      const t = raw.trim()
      if (t) {
        search.append('included_category', t)
      }
    }
  }
  for (const raw of opts?.categoryFiles ?? []) {
    const t = raw.trim()
    if (t) {
      search.append('category_file', t)
    }
  }
  const qs = search.toString()
  const { data } = await apiClient.post<GddNotionSyncRunResponse>(
    `/api/v1/gdd-notion-sync/sync?${qs}`,
    undefined,
    { timeout: API_TIMEOUTS.GDD_NOTION_SYNC },
  )
  return data
}

export async function getGddFullSyncCheckpoint(): Promise<GddFullSyncCheckpointResponse> {
  const { data } = await apiClient.get<GddFullSyncCheckpointResponse>(
    '/api/v1/gdd-notion-sync/full-sync-checkpoint',
  )
  return data
}

export async function deleteGddFullSyncCheckpoint(): Promise<GddFullSyncSimpleOkResponse> {
  const { data } = await apiClient.delete<GddFullSyncSimpleOkResponse>(
    '/api/v1/gdd-notion-sync/full-sync-checkpoint',
  )
  return data
}

export async function postGddFullSyncPause(): Promise<GddFullSyncSimpleOkResponse> {
  const { data } = await apiClient.post<GddFullSyncSimpleOkResponse>(
    '/api/v1/gdd-notion-sync/full-sync/pause',
  )
  return data
}

export async function postGddFullSyncUnpause(): Promise<GddFullSyncSimpleOkResponse> {
  const { data } = await apiClient.post<GddFullSyncSimpleOkResponse>(
    '/api/v1/gdd-notion-sync/full-sync/unpause',
  )
  return data
}

export async function postGddFullSyncCancel(): Promise<GddFullSyncSimpleOkResponse> {
  const { data } = await apiClient.post<GddFullSyncSimpleOkResponse>(
    '/api/v1/gdd-notion-sync/full-sync/cancel',
  )
  return data
}

export interface GddArchiveEntry {
  id: string
  created_at: string
  /** Taille totale des fichiers du snapshot (octets). */
  size_bytes: number
  /** Nombre estimé de fiches GDD (shards + monolithes). */
  fiche_count: number
}

export interface GddArchivesListResponse {
  archives: GddArchiveEntry[]
}

export async function getGddNotionArchives(limit = 20): Promise<GddArchivesListResponse> {
  const { data } = await apiClient.get<GddArchivesListResponse>(
    '/api/v1/gdd-notion-sync/archives',
    { params: { limit } },
  )
  return data
}

export interface GddArchiveRestoreResponse {
  ok: boolean
  message: string
  new_backup_id?: string | null
}

export async function postGddNotionArchiveRestore(
  archiveId: string,
  body: { backup_current?: boolean } = {},
): Promise<GddArchiveRestoreResponse> {
  const { data } = await apiClient.post<GddArchiveRestoreResponse>(
    `/api/v1/gdd-notion-sync/archives/${encodeURIComponent(archiveId)}/restore`,
    { backup_current: body.backup_current ?? true },
  )
  return data
}

export async function getGddNotionSyncStatus(): Promise<GddNotionSyncStatusResponse> {
  const { data } = await apiClient.get<GddNotionSyncStatusResponse>(
    '/api/v1/gdd-notion-sync/status',
  )
  return data
}

export async function getGddNotionSyncProgress(): Promise<GddNotionSyncProgressResponse> {
  const { data } = await apiClient.get<GddNotionSyncProgressResponse>(
    '/api/v1/gdd-notion-sync/sync-progress',
  )
  return data
}

/** Télécharge le ZIP Markdown NotebookLM (sources Notion sync + Vision.json ; max_files côté API défaut 64). */
export type GddNotebooklmExportScope = 'disk' | 'sync'

export async function getGddNotebooklmExportZip(
  maxFiles = 64,
  scope: GddNotebooklmExportScope = 'disk',
): Promise<Blob> {
  const { data } = await apiClient.get<Blob>('/api/v1/gdd-notion-sync/notebooklm-export', {
    params: { max_files: maxFiles, scope },
    responseType: 'blob',
    timeout: API_TIMEOUTS.GDD_NOTEBOOKLM_EXPORT,
  })
  return data
}

File: api/schemas/gdd_notion_sync.py
"""Schémas Pydantic pour la sync GDD Notion (FR18)."""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class GddNotionSourceSchema(BaseModel):
    """Source Notion (page ou base) mappée vers un fichier catégorie GDD."""

    notion_id: str = Field(..., description="UUID page ou base Notion")
    kind: Literal["database", "page"] = Field(..., description='Type de source Notion')
    category_file: str = Field(
        ...,
        description="Nom de fichier cible (ex. personnages.json)",
    )
    notion_data_source_ids: List[str] = Field(
        default_factory=list,
        description=(
            "Optionnel (bases multi-vues Notion) : UUIDs des data sources à interroger. "
            "Vide = sélection automatique (exclut les vues type « récompenses »)."
        ),
    )


class GddNotionSyncConfigPublic(BaseModel):
    """Configuration exposée au client (sans secret)."""

    schema_version: int = 1
    sync_interval_minutes: int = 60
    auto_sync_enabled: bool = False
    sources: List[GddNotionSourceSchema] = Field(default_factory=list)
    included_categories: List[str] = Field(
        default_factory=list,
        description=(
            "Noms de fichier cible (ex. personnages.json) restreignant les sources "
            "kind=database. Vide = toutes les bases et toutes les fiches (page). "
            "Non vide = uniquement ces bases sur ce run ; les fiches page sont exclues."
        ),
    )
    mirror_rebuild_on_full_sync: bool = Field(
        default=False,
        description="Déprécié : ignoré par le serveur (sync complète = miroir automatique).",
    )
    archive_retention_count: int = 10
    token_configured: bool = False


class GddNotionSyncConfigUpdate(BaseModel):
    """Mise à jour partielle de la configuration."""

    sync_interval_minutes: Optional[int] = None
    auto_sync_enabled: Optional[bool] = None
    sources: Optional[List[GddNotionSourceSchema]] = None
    included_categories: Optional[List[str]] = Field(
        default=None,
        description="Filtre bases ; si défini et non vide, les sources page ne sont pas sync sur ce run.",
    )
    mirror_rebuild_on_full_sync: Optional[bool] = Field(
        default=None,
        description="Déprécié : sans effet sur le comportement (conservé pour compat. JSON).",
    )
    archive_retention_count: Optional[int] = Field(
        default=None,
        ge=1,
        description="Nombre max de snapshots sous .archive/ (les plus anciens supprimés).",
    )
    notion_token: Optional[str] = Field(
        default=None,
        description="Si fourni, remplace le token fichier (jamais renvoyé ensuite)",
    )


class GddNotionConnectionTestResponse(BaseModel):
    """Résultat du test de connexion Notion."""

    ok: bool
    message: str
    bot_id: Optional[str] = None
    bot_type: Optional[str] = None


class GddNotionPreviewDatabaseRequest(BaseModel):
    """Corps POST prévisualisation d’une base (première ligne)."""

    category_file: str = Field(..., description="Ex. Dialogues.json — doit exister en source database")


class GddNotionPreviewDatabaseResponse(BaseModel):
    """Résultat prévisualisation : métadonnées Notion + enregistrement mappé comme en sync."""

    ok: bool
    message: str = ""
    category_file: str = ""
    notion_database_id: str = ""
    data_sources_count: int = 0
    data_source_entries: List[Dict[str, str]] = Field(default_factory=list)
    query_total_rows: int = 0
    first_page_id: str = ""
    property_keys_from_query_row: List[str] = Field(default_factory=list)
    property_keys_from_get_page: List[str] = Field(default_factory=list)
    mapped_record: Optional[Dict[str, Any]] = None
    compact_table: bool = False


class GddNotionSyncRunResponse(BaseModel):
    """Réponse après déclenchement sync manuelle."""

    success: bool
    message: str
    updated_entities: int = 0
    partial_errors: List[str] = Field(default_factory=list)


class GddNotionSyncStatusResponse(BaseModel):
    """Statut persisté de la dernière synchronisation."""

    last_started_at: Optional[str] = None
    last_finished_at: Optional[str] = None
    last_success: Optional[bool] = None
    message: str = ""
    updated_entities: int = 0
    partial_errors: List[str] = Field(default_factory=list)
    last_archive_relative: Optional[str] = None
    last_mirror_rebuild_used: Optional[bool] = None


class GddNotionSyncProgressResponse(BaseModel):
    """Progression live d'une sync en cours (polling UI)."""

    active: bool = False
    started_at: Optional[str] = None
    force_full: Optional[bool] = None
    mirror_rebuild: Optional[bool] = None
    phase: str = "idle"
    sources_total: int = 0
    sources_completed: int = 0
    current_source_index: int = 0
    current_category_file: str = ""
    pages_total_known: int = 0
    pages_processed: int = 0
    pages_in_current_source: int = 0
    current_page_in_source: int = 0
    current_page_id_short: str = ""
    message: str = ""
    paused: bool = False


class GddFullSyncCheckpointResponse(BaseModel):
    """État du checkpoint sync complète (reprise possible ou non)."""

    resumable: bool = False
    checkpoint_status: str = Field(
        default="none",
        description="none | resumable | stale | invalid_file",
    )
    checkpoint_file_present: bool = False
    orphan_staging_runs: int = 0
    message: str = ""
    staging_run_name: str = ""
    archive_rel: str = ""
    sources_total: int = 0
    sources_completed: int = 0
    completed_category_files: List[str] = Field(default_factory=list)
    eligible_category_files: List[str] = Field(default_factory=list)


class GddFullSyncCheckpointAbandonResponse(BaseModel):
    """Résultat de l'abandon du checkpoint."""

    ok: bool = True
    message: str = ""


class GddFullSyncPauseResponse(BaseModel):
    """Résultat pause / reprise coopérative."""

    ok: bool = False
    message: str = ""


class GddNotionSyncConfigResponse(BaseModel):
    """Config publique pour GET."""

    config: GddNotionSyncConfigPublic


class GddArchiveEntrySchema(BaseModel):
    """Une entrée d'historique sous ``GDD_categories/.archive/``."""

    id: str = Field(..., description="Nom du dossier snapshot")
    created_at: str = Field(..., description="Horodatage UTC ISO (préfixe du run)")
    size_bytes: int = Field(
        0,
        ge=0,
        description="Taille totale des fichiers du snapshot (octets)",
    )
    fiche_count: int = Field(
        0,
        ge=0,
        description="Nombre estimé de fiches GDD (shards + entrées des monolithes racine)",
    )


class GddArchivesListResponse(BaseModel):
    """Liste des snapshots disponibles."""

    archives: List[GddArchiveEntrySchema] = Field(default_factory=list)


class GddArchiveRestoreRequest(BaseModel):
    """Corps optionnel pour la restauration."""

    backup_current: bool = Field(
        default=True,
        description="Archiver l'état GDD actuel dans .archive/ avant d'appliquer le snapshot.",
    )


class GddArchiveRestoreResponse(BaseModel):
    """Résultat d'une restauration réussie."""

    ok: bool = True
    message: str
    new_backup_id: Optional[str] = Field(
        default=None,
        description="Nom du dossier de sauvegarde créé si backup_current était true",
    )
